examples/python/learning_pytorch.py

```diff
@@ -48,230 +47,250 @@
+
+            if not done:
+                next_state = preprocess(game.get_state().screen_buffer)
+            else:
+                next_state = np.zeros((1, 30, 45)).astype(np.float32)
+
+            agent.append_memory(state, action, reward, next_state, done)
+
+            if global_step > agent.batch_size:
+                agent.train()
+
+            if done:
+                scores.append(game.get_total_reward())
+                train_scores.append(game.get_total_reward())
+                game.new_episode()
+
+            global_step += 1
 
-        self.capacity = capacity
-        self.size = 0
-        self.pos = 0
+        agent.update_target_net()
+        train_scores = np.array(train_scores)
 
-    def add_transition(self, s1, action, s2, isterminal, reward):
-        self.s1[self.pos, 0, :, :] = s1
-        self.a[self.pos] = action
-        if not isterminal:
-            self.s2[self.pos, 0, :, :] = s2
-        self.isterminal[self.pos] = isterminal
-        self.r[self.pos] = reward
+        print("Results: mean: %.1f +/- %.1f," % (train_scores.mean(), train_scores.std()),
+              "min: %.1f," % train_scores.min(), "max: %.1f," % train_scores.max())
 
-        self.pos = (self.pos + 1) % self.capacity
-        self.size = min(self.size + 1, self.capacity)
+        test(game, agent)
+        if save_model:
+            print("Saving the network weigths to:", model_savefile)
+            torch.save(agent.q_net, model_savefile)
+        print("Total elapsed time: %.2f minutes" % ((time() - start_time) / 60.0))
 
-    def get_sample(self, sample_size):
-        i = sample(range(0, self.size), sample_size)
-        return self.s1[i], self.a[i], self.s2[i], self.isterminal[i], self.r[i]
+    game.close()
+    return agent, game
 
 
-class Net(nn.Module):
+class DuelQNet(nn.Module):
     def __init__(self, available_actions_count):
-        super(Net, self).__init__()
-        self.conv1 = nn.Conv2d(1, 8, kernel_size=6, stride=3)
-        self.conv2 = nn.Conv2d(8, 8, kernel_size=3, stride=2)
-        self.fc1 = nn.Linear(192, 128)
-        self.fc2 = nn.Linear(128, available_actions_count)
+        super(DuelQNet, self).__init__()
+        self.conv1 = nn.Sequential(
+            nn.Conv2d(1, 8, kernel_size=6, stride=3, bias=False),
+            nn.BatchNorm2d(8),
+            nn.ReLU()
+        )
+        self.conv2 = nn.Sequential(
+            nn.Conv2d(8, 8, kernel_size=3, stride=2, bias=False),
+            nn.BatchNorm2d(8),
+            nn.ReLU()
+        )
+
+        self.state_fc = nn.Sequential(
+            nn.Linear(96, 64),
+            nn.ReLU(),
+            nn.Linear(64, 1)
+        )
+
+        self.advantage_fc = nn.Sequential(
+            nn.Linear(96, 64),
+            nn.ReLU(),
+            nn.Linear(64, available_actions_count)
+        )
 
     def forward(self, x):
-        x = F.relu(self.conv1(x))
-        x = F.relu(self.conv2(x))
+        x = torch.from_numpy(x).float().to(DEVICE)
```

Comment:
For clarity I would do these transformations outside nn.modules (especially since output is pytorch).

Reply:
Done

Reply:
Done